crypto/crypto.go

```diff
@@ -51,23 +52,33 @@ var (
 
 var errInvalidPubkey = errors.New("invalid secp256k1 public key")
 
+// KeccakState wraps sha3.state. In addition to the usual hash methods, it also supports
+// Read to get a variable amount of data from the hash state. Read is faster than Sum
+// because it doesn't copy the internal state, but also modifies the internal state.
+type KeccakState interface {
+	hash.Hash
+	Read([]byte) (int, error)
+}
+
```

Comment:
Same code in ```trie/hasher.go ```. can they be merged

Reply:
Good point thanks! I merged them so that `trie/hasher.go` uses `crypto.KeccakState` no need for duplication.